app/pages/Recovery/util.ts

```diff
@@ -0,0 +1,304 @@
+ * @param addressBook: the addressBook is used to check for duplicates.
+ */
+export async function addAccounts(
+    accounts: Account[],
+    addressBook: AddressBookEntry[]
+) {
+    for (const account of accounts) {
+        const { address } = account;
+        const accountExists = (await findAccounts({ address })).length > 0;
+        if (!accountExists) {
+            importAccount(account);
+            if (!addressBook.some((abe) => abe.address === address)) {
+                importAddressBookEntry({
+                    readOnly: true,
+                    name: account.name,
+                    address,
+                    note: 'Recovered account',
+                });
+            }
+        }
+    }
+}
+
+/**
+ * Attempts to recover credentials on an identity.
+ * @param prfKeySeed: Seed of the prfKey of the identity.
+ * @param blockHash: block at which the function recover credentials.
+ * @param global: current global parameters.
+ * @param identityId: id of the identity.
+ * @param addressBook: the addressBook is used to check for duplicates when inserting new accounts.
+ * @returns Returns the recovered accounts.
+ */
+export async function recoverFromIdentity(
+    prfKeySeed: string,
+    blockHash: string,
+    global: Global,
+    identityId: number,
+    addressBook: AddressBookEntry[]
+) {
+    const nextCredentialNumber = await getNextCredentialNumber(identityId);
+    const { credentials, accounts } = await recoverCredentials(
+        prfKeySeed,
+        identityId,
+        blockHash,
+        global,
+        nextCredentialNumber
+    );
+
+    if (accounts.length > 0) {
+        await addAccounts(accounts, addressBook);
+    }
+
+    if (credentials.length > 0) {
+        await importCredentials(credentials);
+    }
+    return accounts;
+}
+
+/**
+ * Attempts to recover credentials on an unused identityNumber.
+ * @param prfKeySeed: Seed of the prfKey of the identity.
+ * @param blockHash: block at which the function recover credentials.
+ * @param global: current global parameters.
+ * @param identityNumber: identityNumber of the current wallet to recover from.
+ * @param walletId: id of the wallet to recover from.
+ * @param addressBook: the addressBook is used to check for duplicates when inserting new accounts.
+ * @returns Returns the recovered accounts.
+ */
+export async function recoverNewIdentity(
+    prfKeySeed: string,
+    blockHash: string,
+    global: Global,
+    identityNumber: number,
+    walletId: number,
+    addressBook: AddressBookEntry[]
+) {
+    const { credentials, accounts } = await recoverCredentials(
+        prfKeySeed,
+        0,
+        blockHash,
+        global
+    );
+
+    // If we have found any credentials, create an identity and add the credentials and accounts.
+    // N.B. It is sufficient to check credentials, because accounts are not found without credentials.
+    if (credentials.length) {
+        const identityId = await createRecoveredIdentity(
```

Comment:
The database actions should be transactional, so that we don't end up with an identity with no accounts, or an identity with accounts but no credentials.

Reply:
I have made it all transactionally